File: pyapp/src/pyapp/infrastructure/user_repository.py
from __future__ import annotations
from typing import Optional, Dict, Tuple
import hashlib
import time
import secrets

class UserDto:
    def __init__(self, id: str, email: str, role: str = "user") -> None:
        self.id = id
        self.email = email
        self.role = role

class UserRepository:
    def __init__(self) -> None:
        self._users: Dict[str, Tuple[UserDto, str]] = {}

    def _hash(self, pw: str) -> str:
        return hashlib.sha256(pw.encode("utf-8")).hexdigest()

    async def create(self, email: str, password: str, role: str = "user") -> UserDto:
        uid = f"{int(time.time()*1000)}-{secrets.token_hex(6)}"
        user = UserDto(id=uid, email=email, role=role)
        self._users[email.lower()] = (user, self._hash(password))
        return user

    async def find_by_email(self, email: str) -> Optional[UserDto]:
        rec = self._users.get(email.lower())
        return rec[0] if rec else None

    async def verify(self, email: str, password: str) -> Optional[UserDto]:
        rec = self._users.get(email.lower())
        if not rec:
            return None
        return rec[0] if rec[1] == self._hash(password) else None

user_repo = UserRepository()

File: src/infrastructure/repositories/UserRepository.ts
import { createHash } from 'crypto'
import type { UserDto } from '@/types/user'

/**
 *
 */
export class UserRepository {
    private users = new Map<string, { user: UserDto; passwordHash: string }>()

    private hash(pw: string) {
        return createHash('sha256').update(pw).digest('hex')
    }

    /**
     *
     */
    async create(email: string, password: string, role: UserDto['role'] = 'user') {
        const id = `${Date.now()}-${Math.random().toString(36).slice(2)}`
        const user: UserDto = { id, email, role }
        this.users.set(email.toLowerCase(), { user, passwordHash: this.hash(password) })
        return user
    }

    /**
     *
     */
    async findByEmail(email: string): Promise<UserDto | null> {
        const rec = this.users.get(email.toLowerCase())
        return rec ? rec.user : null
    }

    /**
     *
     */
    async verify(email: string, password: string): Promise<UserDto | null> {
        const rec = this.users.get(email.toLowerCase())
        if (!rec) return null
        const ok = rec.passwordHash === this.hash(password)
        return ok ? rec.user : null
    }
}

export const userRepo = new UserRepository()
